File: 2257-count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: list[list[int]], walls: list[list[int]]) -> int:
        '''
        We first generate the grid and add in those guards and walls
        We go through each cell, 
            - if encounter a G, using DFS, we go all 4 directions and continue to mark all cells that are empty and in the range of this cell to -1 until we hit a W or G
            - if encounter a W, continue

        After marking, we go through each cell again and count those cells that are unmarked
        '''
        grid = [[None] * n for _ in range(m)]

        def mark_guarded(r: int, c: int, dr: int, dc: int):
            nr, nc = r + dr, c + dc
            while 0 <= nr < m and 0 <= nc < n:
                if grid[nr][nc] in ['G', 'W']:
                    break
                grid[nr][nc] = 1
                nr += dr
                nc += dc

        for r, c in guards:
            grid[r][c] = 'G'

        for r, c in walls:
            grid[r][c] = 'W'

        for r in range(m):
            for c in range(n):
                if grid[r][c] == 'G':
                    mark_guarded(r, c, -1, 0)
                    mark_guarded(r, c, 1, 0)
                    mark_guarded(r, c, 0, -1)
                    mark_guarded(r, c, 0, 1)

        res = 0
        for r in range(m):
            for c in range(n):
                if not grid[r][c]:
                    res += 1

        return res


sol = Solution()
m = 4
n = 6
guards = [[0, 0], [1, 1], [2, 3]]
walls = [[0, 1], [2, 2], [1, 4]]
print(f'output: {sol.countUnguarded(m, n, guards, walls)}, expected: 7')
m = 3
n = 3
guards = [[1, 1]]
walls = [[0, 1], [1, 0], [2, 1], [1, 2]]
print(f'output: {sol.countUnguarded(m, n, guards, walls)}, expected: 4')

File: 2257-count-unguarded-cells-in-the-grid.java
class CountUnguardedCellsInTheGrid {
    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {
        int[][] grid = new int[m][n];
        for (int i = 0; i < guards.length; i++) {
            int r = guards[i][0], c = guards[i][1];
            grid[r][c] = 'G';
        }
        for (int i = 0; i < walls.length; i++) {
            int r = walls[i][0], c = walls[i][1];
            grid[r][c] = 'W';
        }

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (grid[r][c] == 'G') {
                    markGuarded(grid, r, c, -1, 0, m, n);
                    markGuarded(grid, r, c, 1, 0, m, n);
                    markGuarded(grid, r, c, 0, -1, m, n);
                    markGuarded(grid, r, c, 0, 1, m, n);
                }
            }
        }

        int res = 0;

        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                if (grid[r][c] == 0) {
                    res += 1;
                }
            }
        }

        return res;
    }

    private void markGuarded(int[][] grid, int r, int c, int dr, int dc, int m, int n) {
        int nr = r + dr, nc = c + dc;
        while (nr >= 0 && nr < m && nc >= 0 && nc < n) {
            if (grid[nr][nc] == 'G' || grid[nr][nc] == 'W') {
                break;
            }
            grid[nr][nc] = 1;
            nr += dr;
            nc += dc;
        }
    }

    public static void main(String[] args) {
        CountUnguardedCellsInTheGrid sol = new CountUnguardedCellsInTheGrid();
        System.out.println("Output: " + sol.countUnguarded(4, 6, new int[][] { { 0, 0 }, { 1, 1 }, { 2, 3 } },
                new int[][] { { 0, 1 }, { 2, 2 }, { 1, 4 } }) + ", expected: 7");
        System.out.println("Output: " + sol.countUnguarded(3, 3, new int[][] { { 1, 1 } },
                new int[][] { { 0, 1 }, { 1, 0 }, { 2, 1 }, { 1, 2 } }) + ", expected: 4");
    }
}
